### ISSUE TITLE
`DirectiveLocator:associated()` performance 

### ISSUE THREAD

**Issue by LastDragon-ru** [nuwave/lighthouse#2041]
We intensively use directives and some of our queries return hundreds and thousands of objects. I've found that each call of `DirectiveLocator::associated($node)` create new instances of directives and this lead to (for 500 objects)

![image](https://user-images.githubusercontent.com/1329824/151311897-becc2394-052b-4838-a9dc-a5e6948a0456.png)

In PHP 8.0 it is possible to use `WeakMap` to cache directive instances and we will get:

![image](https://user-images.githubusercontent.com/1329824/151312027-1799d346-96c0-4a21-bf04-aed768b42de8.png)

```php
<?php declare(strict_types = 1);

namespace App\GraphQL\Extensions\Lighthouse;

use GraphQL\Language\AST\Node;
use Illuminate\Contracts\Events\Dispatcher;
use Illuminate\Support\Collection;
use Nuwave\Lighthouse\Schema\DirectiveLocator as LighthouseDirectiveLocator;
use WeakMap;

class DirectiveLocator extends LighthouseDirectiveLocator {
    /**
     * @var \WeakMap<\GraphQL\Language\AST\Node,\Illuminate\Support\Collection<\Nuwave\Lighthouse\Support\Contracts\Directive>
     */
    private WeakMap $directives;

    public function __construct(Dispatcher $eventsDispatcher) {
        parent::__construct($eventsDispatcher);

        $this->directives = new WeakMap();
    }

    public function associated(Node $node): Collection {
        if (!isset($this->directives[$node])) {
            $this->directives[$node] = parent::associated($node);
        }

        return $this->directives[$node];
    }
}
```

**Comment by LastDragon-ru** [comment]
I've found that cache should be disabled while the AST Manipulation phase (because directives/args can change), so the first post updated.

**Comment by LastDragon-ru** [comment]
I've also found that `ValidatorDirective` doesn't not call `Validator::setArgs()` after `ValidatorDirective::setArgumentValue()` call that is critical when directives cached. The first post updated.

**Comment by spawnia** [comment]
This change has the potential for a subtle breakage, exactly due to directives such as `ValidatorDirective` that have lingering state. The same might be true for custom directives.

The PHP compatiblity can be resolved by using `SplObjectStorage`. I would be happy about a pull request so I can benchmark this for our application.

**Comment by LastDragon-ru** [comment]
> This change has the potential for a subtle breakage, exactly due to directives such as ValidatorDirective

Yep. Also, a bit related issue: `BaseDirective::hydrate()` doesn't reset cached args.

> I would be happy about a pull request so I can benchmark this for our application.

In our real/live application, we have improvements of about ~6%. I will try to create PR next week, but no promises.
